source/vsignal/sink.d

```diff
@@ -0,0 +1,199 @@
+module vsignal.sink;
+
+import vsignal.signal;
+import vsignal.slot : Slot, slot_connect = connect;
+
+/**
+Connect a listener to a Signal.
+
+A payload can be passed if the function to call is a data member function of the
+same or if the function accepts as its first parameter a reference to its type.
+
+Params:
+	pred = the listener to connect.
+	sink = the Sink that holds the signal.
+	instance = the payload to store.
+
+Returns: A Connection of the listener.
+*/
+Connection connect(alias pred, F)(auto ref Sink!F sink)
+{
+	sink.disconnect!pred;
+
+	import core.lifetime : move;
+
+	with (sink)
+	{
+		Slot!F call;
+		call.slot_connect!pred;
+		signal.calls ~= call.move();
+
+		Slot!(void delegate(void* signal) @safe pure nothrow) conn;
+		conn.slot_connect!(release!(pred, F))();
+
+		return Connection(conn.move(), signal);
+	}
+}
+
+///
+Connection connect(alias pred, T, F)(auto ref Sink!F sink, ref T instance)
+{
+	sink.disconnect!pred(instance);
+
+	import core.lifetime : move;
+
+	with (sink)
+	{
+		Slot!F call;
+		call.slot_connect!pred(instance);
+		signal.calls ~= call.move();
+
+		Slot!(void delegate(void* signal) @safe pure nothrow) conn;
+		conn.slot_connect!(release!(pred, T, F))(instance);
+
+		return Connection(conn.move(), signal);
+	}
+}
+
+/++
+Disconnects a listener from a Signal. The order is not preserved after the
+removal.
+
+Examples:
+---
+void method() {}
+struct Foo { void method() {} }
+Foo foo;
+
+/* ... */
+
+// lets assume a signal exists with listeners
+sink.disconnect!(Foo.method)(foo); // disconnects Foo.method with payload foo
+sink.disconnect!method;            // disconnects method
+sink.disconnect(foo);              // disconnects all listeners with payload foo
+sink.disconnect();                 // disconnects all listeners
+---
+
+Params:
+	pred = the listener to disconnect.
+	sink = the Sink that holds the signal.
+	instance = the payload that composes the listener.
++/
+void disconnect(alias pred, F)(auto ref Sink!F sink)
+{
+	import std.algorithm : remove, SwapStrategy;
+	import std.algorithm : countUntil;
+
+	Slot!F call;
+	call.slot_connect!pred;
+
+	with (sink)
+	{
+		const index = signal.calls.countUntil(call);
```

Comment:
For readability I would specify what type this is.

Reply:
The variable is temporary and its name documents it already. I don't think specifying `const size_t` makes a difference in readability.